src/hotspot/share/nmt/nmtTreap.hpp

```diff
@@ -0,0 +1,389 @@
+    assert(recur_count < 200, "Call-stack depth should never exceed 200");
+
+    if (left == nullptr) return right;
+    if (right == nullptr) return left;
+
+    if (left->_priority > right->_priority) {
+      // We need
+      //      LEFT
+      //         |
+      //         RIGHT
+      // for the invariant re: priorities to hold.
+      left->_right = merge(left->_right, right DEBUG_ONLY(COMMA recur_count + 1));
+      return left;
+    } else {
+      // We need
+      //         RIGHT
+      //         |
+      //      LEFT
+      // for the invariant re: priorities to hold.
+      right->_left = merge(left, right->_left DEBUG_ONLY(COMMA recur_count + 1));
+      return right;
+    }
+  }
+
+  static TreapNode* find(TreapNode* node, const K& k DEBUG_ONLY(COMMA int recur_count = 0)) {
+    if (node == nullptr) {
+      return nullptr;
+    }
+
+    if (COMPARATOR::cmp(node->_key, k) == 0) { // EQ
+      return node;
+    }
+
+    if (COMPARATOR::cmp(node->_key, k) <= 0) { // LEQ
+      return find(node->_left, k DEBUG_ONLY(COMMA recur_count + 1));
+    } else {
+      return find(node->_right, k DEBUG_ONLY(COMMA recur_count + 1));
+    }
+  }
+
+#ifdef ASSERT
+  bool verify_self() {
+    double expected_maximum_depth = log(this->_node_count+1) * 5;
+    // Find the maximum depth through DFS and ensure that the priority invariant holds.
+    int maximum_depth_found = 0;
+
+    struct DFS {
+      int depth;
+      uint64_t parent_prio;
+      TreapNode* n;
+    };
+    GrowableArrayCHeap<DFS, mtNMT> to_visit;
+    uint64_t positive_infinity = 0xFFFFFFFFFFFFFFFF;
+
+    to_visit.push({0, positive_infinity, this->_root});
+    while (!to_visit.is_empty()) {
+      DFS head = to_visit.pop();
+      if (head.n == nullptr) continue;
+      if (maximum_depth_found < head.depth) {
+        maximum_depth_found = head.depth;
+      }
+      if (head.parent_prio < head.n->_priority) {
+        return false;
+      }
+      to_visit.push({head.depth + 1, head.n->_priority, head.n->left()});
+      to_visit.push({head.depth + 1, head.n->_priority, head.n->right()});
+    }
+    if (maximum_depth_found > (int)expected_maximum_depth) {
+      return false;
+    }
+    // Visit everything in order, see that the key ordering is monotonically increasing.
+    TreapNode* last_seen = nullptr;
+    bool failed = false;
+    this->visit_in_order([&](TreapNode* node) {
+      if (last_seen == nullptr) {
+        last_seen = node;
+        return;
+      }
+      int c = COMPARATOR::cmp(last_seen->key(), node->key());
+      if (c > 0) {
+        failed = false;
+      }
+      last_seen = node;
+    });
+    return !failed;
+  }
+#endif // ASSERT
+
+public:
+  Treap(uint64_t seed = static_cast<uint64_t>(os::random()))
+  : _root(nullptr),
+  _prng_seed(seed)
+  DEBUG_ONLY(COMMA _node_count(0)) {
+  }
+
+  ~Treap() {
+    this->remove_all();
+  }
+
+  void upsert(const K& k, const V& v) {
+    assert(verify_self(), "invariant");
```

Comment:
Here, and down in remove(): I would consider modifying the verification here, or even remove it, and call it explicitly by callers of the treap at convenient times only. My experiences from doing very similar things in Metaspace was that such fine grained verifications get quickly way to expensive, even for a debug build.

In metaspace, I ended up doing something like this: https://github.com/openjdk/jdk/blob/ac86f59e4f5382d5c3e8984532dd210611db7dcb/src/hotspot/share/memory/metaspace/metaspaceCommon.hpp#L119

which adds assertion that are only checked every n times, in that case configurable with a switch.